driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -11,19 +15,77 @@ use futures::{
 };
 use orderbook::Orderbook;
 use std::future::Future;
-use std::sync::{Arc, Mutex};
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc, Mutex,
+};
+
+/// An event based orderbook that automatically updates itself with new events from the contract.
+#[derive(Debug)]
+pub struct UpdatingOrderbook {
+    orderbook: Arc<Mutex<Orderbook>>,
+    orderbook_ready: Arc<AtomicBool>,
+    // When this struct is dropped this sender will be dropped which makes the updater thread stop.
+    _exit_tx: oneshot::Sender<()>,
+}
+
+impl UpdatingOrderbook {
+    pub fn new(
+        _contract: &impl StableXContract,
+        block_timestamp: impl BlockTimestamp + Send + 'static,
+    ) -> Self {
+        let orderbook = Arc::new(Mutex::new(Orderbook::default()));
+        let orderbook_clone = orderbook.clone();
+        let orderbook_ready = Arc::new(AtomicBool::new(false));
+        let orderbook_ready_clone = orderbook_ready.clone();
+        let (exit_tx, exit_rx) = oneshot::channel();
+        // Create stream first to make sure we do not miss any events between it and past events.
+        // TODO: use the real functions once they are implemented
+        let stream = futures::stream::iter(vec![]).boxed(); // contract.stream_events();
+        let past_events = futures::future::ready(Ok(Vec::new())).boxed(); // contract.past_events();
+
+        std::thread::spawn(move || {
+            let result = futures::executor::block_on(update_with_events_forever(
+                orderbook_clone,
+                orderbook_ready_clone,
+                block_timestamp,
+                exit_rx,
+                past_events,
+                stream,
+            ));
+            if let Err(err) = result {
+                log::error!("event based orderbook failed: {}", err);
+            }
```

Comment:
I am wondering about this. Let's say the node goes down for a short time. In this version the orderbook would fail and never update again but the program as a whole wouldn't exit. We should probably add a retry mechanism to recreate the orderbook from scratch if `update_with_events_forever` exits.

Reply:
yes, we should do that. For now we could also panic, which would cause the pod to restart, but an intrinsic retry mechanism is probably good going forward.
I wonder if the stream itself could become a retrying stream (since this could avoid refetching the entire orderbook again).

Reply:
I added an `abort`. Panic isn't enough because this happens in a thread.